onnxruntime/core/graph/graph_utils.cc

```diff
@@ -422,6 +422,22 @@ size_t RemoveNodeOutputEdges(Graph& graph, Node& node) {
   return output_edges.size();
 }
 
+void BuildTensorProtoForInitializer(const Tensor& out_tensor, const NodeArg& node_arg,
+                                    ONNX_NAMESPACE::TensorProto& tensor_proto) {
+  // Set name, dimensions, type, and data of the TensorProto.
+  tensor_proto.set_name(node_arg.Name());
+
+  for (auto& dim : out_tensor.Shape().GetDims()) {
+    tensor_proto.add_dims(dim);
+  }
+  auto tensorproto_type = node_arg.TypeAsProto()->tensor_type().elem_type();
```

Comment:
>TypeAsProto() [](start = 35, length = 13)

Could you just init tensor_proto by copying from node_arg.TypeAsProto()->tensor_type() and add the data after that?

Reply:
There is only a constructor with no parameters for TensorProto. 
I will simplify a bit the code to avoid a couple of temp variables.

Reply:
What if the has external data?

Reply:
Why do you need the node_arg argument here? Should a Tensor and string name are enough?

Reply:
We also need the node_arg to get the element type. Is there a way to get that from the Tensor?

Reply:
I just tried using `tensor.DataType()->GetTypeProto()->tensor_type().elem_type()`, but the data type of the tensor is not always known.
Maybe I can pass a name and a type instead of a node_arg, it might be cleaner. Then it's on the caller to get the type our of the node_arg.

Reply:
You can get the info from tensor.DataType(), which returns a pointer, you can use the 'GetTensorProtoType' function in core\framework\tensorprotoutils.cc to cast it to a enum.

I'd prefer to have a function which is like: 

```c++
ONNX_NAMESPACE::TensorProto SerializeTensorToProto(const Tensor& tensor, const std::string* optional_name);
```

It's more generic and easier to test.

Reply:
As discussed offline, the current GetTensorProtoType() is missing a few data types, so for now I will pass the data type as argument. Once we support all data types, we can remove the argument and get it from the tensor directly.
Also, given that the method became generic enough, I will move it to the tensorprotoutils, where we can also check endianness.